olive/passes/onnx/mnb_to_qdq.py

```diff
@@ -0,0 +1,307 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+import math
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict
+
+import numpy as np
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import get_external_data_config, model_proto_to_olive_model
+from olive.passes.onnx.onnx_dag import OnnxDAG
+from olive.passes.pass_config import PassConfigParam
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+
+logger = logging.getLogger(__name__)
+
+
+class MatMulNBitsToQDQ(Pass):
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "use_transpose_op": PassConfigParam(
+                type_=bool,
+                # TODO(jambayk): decide whether to enable this by default or not
+                # False gives same output on arm Mac/Windows, but not on x64 Linux/Windows
+                default_value=True,
+                description=(
+                    "Whether to use a Transpose operator after the DequantizeLinear operator. If False, the weight"
+                    " initializer will be transposed instead."
```

Comment:
Add a note in the description regarding in which scenarios user may want to set this to false.

Reply:
I updated the default value to be True since it is more standard to just have DQ->MatMul. Added a note that using transpose op might be more efficient for eps such as DirectML. For other EPs, I am not sure which one is better since the int4 support is new and might not be optimized either way.